search/localsearch_nqueens.cpp

```diff
@@ -0,0 +1,186 @@
+ * The rule of neighborhood applied in this file is first.improvement.
+ * @author [dhernandezl](https://github.com/dhernandezl)
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <ctime>
+
+#define NREINAS 8 //The number of queens on the table is the same as the dimension of the chessboard.
+				  // NREINAS 4; NREINAS 8; NREINAS 16
+/**
+ * function: initial_solution
+ * Initialize a new solution, generating
+ * random numbers with the maximum height of the board.
+ * @returns a vector that will be the starting solution.
+ */
+std::vector<int> initial_solution() {
+	std::vector<int> solution;
+	srand(time(nullptr)); //reset srand
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		solution.push_back(rand() % NREINAS + 1);
+	}
+	return solution;
+}
+
+/**
+ * function: cost
+ * Calculate the cost of the solution vector,
+ * the optimal cost is h = 1. Read the 8 queens problem.
+ * @param [in] solution_vector is a vector with the
+ * column number where the queens are located on the board.
+ * @returns the cost of the solution vector.
+ */
+int cost(std::vector<int> solution_vector) {
+	int h = 0;
+	std::vector<int> aux_sol;
+	//sum column
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (solution_vector[i] == solution_vector[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	//sum diagonal
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		aux_sol.push_back(solution_vector[i] - (i + 1));
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (aux_sol[i] == aux_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	aux_sol.clear();
+	//sum inverse diagonal
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		aux_sol.push_back(solution_vector[i] + (i + 1));
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (aux_sol[i] == aux_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	return h;
+}
+
+/**
+ * function: neighbour
+ * Generate a neighbor based on the current solution of
+ * the problem. Read about neighborhoods in local search.
+ * @param [in] solution_vector is a vector with the
+ * column number where the queens are located on the board.
+ * @returns a new vector similar to the one passed by parameter.
+ */
+std::vector<int> neighbour(std::vector<int> solution_vector) {
+	srand(time(nullptr)); //reset srand
+	int i = rand() % NREINAS + 1;
+	srand(time(nullptr)); //reset srand
+	int mov = (rand() % NREINAS + 1) - i;
+	mov = (mov < 0) ? mov *= -1 : mov;
+	mov = (mov == 0) ? mov = 1 : mov;
+	solution_vector[i - 1] = mov;
+	return solution_vector;
+}
+
+/**
+ * function: print_solution
+ * Prints the solution obtained on the chessboard in console.
+ * @param [in] solution_vector is a vector with the
+ * column number where the queens are located on the board.
+ */
+void print_solution(std::vector<int> solution_vector) {
+	char chessboard[NREINAS][NREINAS] = {};

```

Comment:
Don't use C-style arrays in a C++ program.
These kind of arrays cause errors and are not good.

Use `std::array<>` or `std::vector<>` instead.

```suggestion
	std::vector<std::vector<char>> chessboard = {};
```